src/java.base/share/classes/java/lang/StringLatin1.java

```diff
@@ -326,30 +433,72 @@ private static int indexOfChar(byte[] value, int ch, int fromIndex, int max) {
     }
 
-    @IntrinsicCandidate
+    /**
+     * Searches for the first occurrence of {@code str} in {@code value}, and,
+     * if found, returns the index of the first character in {@code value};
+     * {@code -1} otherwise.
+     *
+     * @param value a Latin-1 string byte array to search in
+     * @param str a Latin-1 string byte array to search for
+     * @return the index of the first character in {@code value} if a match is
+     * found; {@code -1} otherwise
+     *
+     * @throws NullPointerException if {@code value} or {@code str} is null
+     */
     static int indexOf(byte[] value, byte[] str) {
+        Objects.requireNonNull(value);
+        Objects.requireNonNull(str);
+        return indexOf0(value, str);
+    }
+
+    // inline_string_indexOf
+    @IntrinsicCandidate
+    private static int indexOf0(byte[] value, byte[] str) {
         if (str.length == 0) {
             return 0;
         }
         if (value.length == 0) {
             return -1;
         }
-        return indexOf(value, value.length, str, str.length, 0);
+        return indexOf0(value, value.length, str, str.length, 0);
     }
 
+    /**
+     * Searches for the first occurrence of the given {@code str} sub-range in
+     * the given {@code value} sub-range, and, if found, returns the index of
+     * the first character in {@code value}; {@code -1} otherwise.
+     *
+     * @param value a Latin-1 string byte array to search in
+     * @param valueToIndex the index (exclusive) of the last character in {@code value}
+     * @param str a Latin-1 string byte array to search for
+     * @param strToIndex the index (exclusive) of the last character in {@code str}
+     * @param valueFromIndex the index (inclusive) of the first character in {@code value}
+     * @return the index of the first character in {@code value} if a match is
+     * found; {@code -1} otherwise
+     *
+     * @throws NullPointerException if {@code value} or {@code str} is null
+     * @throws StringIndexOutOfBoundsException if the sub-ranges are out of bounds
+     */
+    static int indexOf(byte[] value, int valueToIndex, byte[] str, int strToIndex, int valueFromIndex) {
+        String.checkBoundsBeginEnd(valueFromIndex, valueToIndex, value.length);    // Implicit null check on `value`
+        String.checkBoundsBeginEnd(0, strToIndex, str.length);                     // Implicit null check on `str`
+        return indexOf0(value, valueToIndex, str, strToIndex, valueFromIndex);
+    }
+
+    // inline_string_indexOfI
     @IntrinsicCandidate
-    static int indexOf(byte[] value, int valueCount, byte[] str, int strCount, int fromIndex) {
+    private static int indexOf0(byte[] value, int valueToIndex, byte[] str, int strToIndex, int valueFromIndex) {
         byte first = str[0];
-        int max = (valueCount - strCount);
-        for (int i = fromIndex; i <= max; i++) {
+        int max = (valueToIndex - strToIndex);
+        for (int i = valueFromIndex; i <= max; i++) {
```

Comment:
Note that the earlier code used `valueCount` – the same `Count`-suffixed variables are used at the intrinsics too. Though as can be seen from these two lines, what is meant is a `ToIndex`. That is, `valueCount` does *not* denote _"the count of characters"_ after the from index, but instead it denotes _"the (exclusive) index of the last character"_. I've updated the Java code to reflect this, though I've not touched the intrinsic.

Note that the same issue (and the correction) is present in the `StringUTF16::indexOf` family too.